user_data/bin/check_max_conn.escript

```diff
@@ -0,0 +1,122 @@
+
+%%% Client entry
+main(["-c", Host, Port0, Parallel]) ->
+    main(["-c", Host, Port0, Parallel, ""]);
+
+main(["-c", Host, Port0, Parallel, AddrsStr]) ->
+    process_flag(trap_exit, true),
+    LocalAddrs0 = lists:map(
+                    fun(A) ->
+                            {ok, Addr} = inet_parse:address(A),
+                            Addr
+                    end, string:tokens(AddrsStr, ",")),
+    LocalAddrs = case LocalAddrs0 of
+                     [] -> [{0,0,0,0}];
+                     _ -> LocalAddrs0
+                 end,
+
+    NumberOfClients = list_to_integer(Parallel),
+    Port = list_to_integer(Port0),
+    ok = persistent_term:put(counter, counters:new(3, [write_concurrency])),
+    lists:foreach(
+      fun(X) ->
+              spawn_link(fun() -> run_client(Host, Port, X, LocalAddrs) end)
+      end, lists:seq(1, NumberOfClients)),
+    main_loop(NumberOfClients);
+main(_) ->
+    io:format("Check MAX concurrent TCP connections between server/client peers \n\n\n"
+              "* Start on Server side: \n"
+              "  check_max_conn.escript -s Port ParallelWorkers\n"
+              "  e.g.~n"
+              "  check_max_conn.escript -s 1883 256 \n"
+              "\n"
+              "* Start on Client side: rate per worker is 100/s \n"
+              "  check_max_conn.escript -c Host Port ParallelWorkers [LocalAddrs]\n"
+              "  e.g.~n"
+              "  check_max_conn.escript -c lb.int.emqx 1883 10\n"
+              "  = for more than 64k conns, multiple local ips is required=~n"
+              "  check_max_conn.escript -c lb.int.emqx 1883 10 192.168.1.2,192.168.1.3\n"
+             ).
+
+%% Client
+run_client(Host, Port, Id, Addrs) ->
+    run_client(Host, Port, Id, Addrs, 1).
+
+run_client(Host, Port, Id, Addrs, Offset) when Offset > length(Addrs) ->
+    run_client(Host, Port, Id, Addrs, 1);
+run_client(Host, Port, Id, Addrs, Offset) ->
+    Addr = lists:nth(Offset, Addrs),
+    Cnt = persistent_term:get(counter),
+    case gen_tcp:connect(Host, Port, [{ip, Addr}]) of
+        {ok, _Port} ->
+            ok;
+        {error, _} = E ->
+            io:format("Error ~p to connect with local addr ~p~n", [E, Addr]),
+            exit({conn_error, E})
+    end,
+    ok = counters:add(Cnt, 1, 1),
+    receive
+        {tcp_closed, _Socket} ->
+            ok = counters:add(Cnt, 2, 1);
+        {tcp_error, _Socket, _Reason} ->
+            ok = counters:add(Cnt, 3, 1)
+    after 10 -> %% 100 conns/sec
+            ok
+    end,
+    run_client(Host, Port, Id, Addrs, Offset+1).
+
+%% Server Acceptors
+acceptor_loop(LPort, Id) ->
+    {ok, _P } = gen_tcp:accept(LPort),
+    Cnt = persistent_term:get(counter),
+    ok = counters:add(Cnt, 1, 1),
+    receive
+        {tcp_closed, _Socket} ->
+            ok = counters:add(Cnt, 2, 1);
+        {tcp_error, _Socket, _Reason} ->
+            ok = counters:add(Cnt, 3, 1)
+    after 0 ->
+            ok
+    end,
+    acceptor_loop(LPort, Id).
+
+main_loop(0) ->
```

Comment:
Q: reaching this should be rare, right?  Since this will only happen if all clients/acceptors die?

Reply:
right, this is mostly to deal with incorrect user inputs such as non-existing src ips ...